File: Labs/Lab4/lab4-pendulum_RK2.py
#
# damped, driven nonlinear pendulum motion
#
# ODE system:   1) dtheta/dt = omega
#               2) domega/dt = -Omega0^2 * sin(theta) - q omega + fD sin(OmegaD t)
#
#     where Omega0^2 = g / L
#
# solved numerically via the 2nd-order Runge-Kutta method
#
#
# by Denes Molnar for PHYS 580, Fall 2020


from math import sin, cos, sqrt, floor
from math import pi as PI
import numpy as np

import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt


# alternatively, you could use this function to limit any angle to [-pi,pi)
def chopAngle(th0):
   theta = th0 - floor(th0 / (2. * PI)) * 2. * PI    # chop to [0, 2pi)
   if theta >= PI: theta -= 2. * PI                  # shift to [-pi,pi)
   return theta


# compute evolution of theta and omega from given initconds
def computeTrajectory(initconds, params, dt, nsteps):
   (th0, om0) = initconds
   (Omega0, q, fD, OmegaD) = params

   Omega0sq = Omega0**2

   # allocate storage - store both theta and omega
   theta = np.zeros(nsteps + 1)
   omega = np.zeros(nsteps + 1)
   theta[0], omega[0] = th0, om0

   # solve using 2nd-order Runge-Kutta method
   t = 0
   for i in range(nsteps):
      # halfstep
      om_half = omega[i] + 0.5 * dt * ( -Omega0sq * sin(theta[i]) - q * omega[i] + fD * sin(OmegaD * t) )
      th_half = theta[i] + 0.5 * dt * omega[i]
      t_half = t + 0.5 * dt
      # full step
      omega[i + 1] = omega[i] + dt * ( -Omega0sq * sin(th_half) - q * om_half + fD * sin(OmegaD * t_half) )
      theta[i + 1] = theta[i] + dt * om_half
      t += dt
      # bound theta (or, you could use the chop function -> slower)
      if   theta[i + 1] >= PI:  theta[i + 1] -= 2. * PI
      elif theta[i + 1] < -PI:  theta[i + 1] += 2. * PI 

   # return results
   return theta, omega


# get parameters
#

# physical parameters and initial conditions (units are in [] brackets)
L   = float( input('length of pendulum [m]: ') )
q   = float( input('dissipation coefficient [1/s]: ') )
fD  = float( input('driving coefficient [1/s^2]: ') )
OmegaD = float( input('driving angular frequency [rad/s]: ') )
th0  = float( input('initial angle wrt to vertical [rad]: ') )
om0  = float( input('initial angular velocity [rad/s]: ') )
#
g    = 9.8      # gravitational acceleration [m/s^2] (hardcoded)

Omega0 = sqrt(g / L)

# computational parameters: dt[s], nsteps
dt = float( input('dt [s]: ') )
nsteps = int( input('number of time steps to calculate: ') )


# compute evolution
#
params = (Omega0, q, fD, OmegaD)
initconds = (th0, om0)

theta, omega   = computeTrajectory(initconds, params, dt, nsteps)

# time values, for plotting
timevals = np.array( [ dt * i  for i in range(nsteps + 1) ] )
# calculate 'energy' at each timestep 
energy  = 0.5 * omega**2  + Omega0**2 * (1. - np.cos(theta))   # use numpy vector ops


# plots
#

#fontsize = 30   # fontsize, need to be tuned to screen resolution
fontsize = 15    # fontsize, for 1920x1080
linewidth = 3
pointarea = 2.

# make a 2x2 array of plots 
fig, axArray = plt.subplots(2, 2)

((ax0, ax1), (ax2, ax3)) = axArray

# theta-omega  (TOP LEFT)
#ax0.plot(theta, omega, linewidth = linewidth, label = "phase space")
ax0.scatter(theta, omega, s = pointarea, label = "phase space")

ax0.tick_params(labelsize = fontsize)
ax0.legend(fontsize = fontsize)
ax0.set_xlabel("theta [rad]", fontsize = fontsize)
ax0.set_ylabel("omega [rad/s]", fontsize = fontsize)

titleString  = "dt=" + str(dt) + ", L=" + str(L) 
titleString += ", fD=" + str(fD) + ", OmegaD=" + str(round(OmegaD, 3))
titleString += ", omega0=" + str(om0) + ", theta0=" + str(th0)
ax0.set_title(titleString, fontsize = fontsize)

# omega (TOP RIGHT)
ax1.plot(timevals, omega, linewidth = linewidth, label = "omega")

ax1.tick_params(labelsize = fontsize)
ax1.legend(fontsize = fontsize)
ax1.set_xlabel("t[s]", fontsize = fontsize)
ax1.set_ylabel("omega[rad/s]", fontsize = fontsize)

# theta (BOTTOM LEFT)
ax2.plot(timevals, theta, linewidth = linewidth, label = "theta")

ax2.tick_params(labelsize = fontsize)
ax2.legend(fontsize = fontsize)
ax2.set_xlabel("t[s]", fontsize = fontsize)
ax2.set_ylabel("theta[rad]", fontsize = fontsize)


# energy (BOTTOM RIGHT)
ax3.plot(timevals, energy, linewidth = linewidth, label = "energy")

ax3.tick_params(labelsize = fontsize)
ax3.legend(fontsize = fontsize)
ax3.set_xlabel("t[s]", fontsize = fontsize)
ax3.set_ylabel("energy [a.u.]", fontsize = fontsize)


# show plot
plt.subplots_adjust(wspace = 0.4, hspace = 0.2)
#fig.tight_layout()
plt.show()


#EOF

File: Labs/Lab04/Lab4-pendulumRK2.m
%
% RK2 calculation for the motion of a non-linear pendulum
%
% by Kevin Berwick, edited slightly by H. Nakanishi and D. Molnar
% based on 'Computational Physics' book by N Giordano and H Nakanishi,
% Section 3.1
%
clear all;
%
% Initialize physical parameters
% - ordinary pendulum
length = input('Length of pendulum [m]: ');
theta0 = input('Initial angle [rad]: ');
omega0 = input('initial angular velocity [rad/s]: ');
% - dissipation and driving force
q      = input('dissipation coefficient [1/s]: ');
fD     = input('driving coefficient [1/s^2]: ');
OmegaD = input('driving angular frequency [rad/s]: ');
%
g = 9.8;
Omega0sq = g / length;  % natural ang. frequency squared
%
% Initialize computational parameters
%
dt     = input('dt [s]: ');
nsteps = input('number of steps: ');
%
% Initialize arrays for time, angle, angular velocity
% and set initial conditions
%
time   = zeros(nsteps+1,1);
theta  = zeros(nsteps+1,1);
omega  = zeros(nsteps+1,1);
% we could track E, but it can be reconstructed at end from theta & omega
%energy = zeros(nsteps+1,1);
%
theta(1) = theta0;
omega(1) = omega0;
%energy(1) = 0.5 * length^2 * omega0^2 + g * length * (1 - cos(theta0));
%
% Loop over the timesteps using the RK2 approximaton
%
for i = 1:nsteps
  % RK2 half step
	omp = omega(i) + 0.5 * dt * ( - Omega0sq * sin(theta(i)) - q * omega(i) ...
		                            + fD * sin(OmegaD*time(i)) );
  thp = theta(i) + 0.5 * dt * omega(i);
	tp  = time(i)  + 0.5 * dt;
  % RK2 full step
	omega(i+1) = omega(i) + dt * ( - Omega0sq * sin(thp) - q * omp ...
                                 + fD * sin(OmegaD*tp) );
  th1 = theta(i) + omp * dt;
  % cut theta modulo 2pi, map to [-pi,pi] range, store constrained value
  % -> floor(x) returns highest integer that is <= x
  th1 = th1 - floor(th1 / (2*pi)) * 2*pi;
  if (th1 > pi)
    th1 = th1 - 2*pi;
  end;
  theta(i+1) = th1;
  %energy(step+1) = 0.5*length^2*omega(step+1)^2+g*length*(1-cos(theta(step+1)));
	time(i+1) = time(i) + dt;
end;
%
% Show results as a 2x2 array of plots
%
% reconstruct energy
energy = 0.5 * length^2 * omega.^2 + g * length * (1 - cos(theta));
% plots
subplot(2, 2, 1);
plot(time, theta, 'r', 'LineWidth', 3);
legend('Angle');
xlabel('time [s]');
ylabel('theta [rad]');
%
subplot(2, 2, 2);
plot(time, omega, 'b', 'LineWidth', 3);
legend('Angular velocity');
xlabel('time [s]');
ylabel('omega [rad/s]');
%
subplot(2, 2, 3);
plot(theta, omega, 'm', 'LineWidth', 3);
%plot(omega, energy, 'm', 'LineWidth', 3);
legend('Phase portrait');
xlabel('theta [rad]');
ylabel('omega [rad/s]');
title( ['RK2: dt = ', num2str(dt), ', th(0) = ', num2str(theta0), ...
        ', om(0) = ', num2str(omega0), ', l = ', num2str(length)] )
%
subplot(2, 2, 4);
plot(time, energy, 'g', 'LineWidth', 3);
legend('Energy');
%ylim([0,2*energy(1)]);
xlabel('time [s]');
ylabel('energy/mass [J/kg]');
title([', q = ', num2str(q), ', f_D = ', num2str(fD), ...
       ', OmegaD = ', num2str(OmegaD)])
